pubsub/redis/redis.go

```diff
@@ -113,68 +180,215 @@ func (r *redisStreams) Publish(req *pubsub.PublishRequest) error {
+	for _, msg := range msgs {
+		var data []byte
+		if dataValue, exists := msg.Values["data"]; exists && dataValue != nil {
+			switch v := dataValue.(type) {
+			case string:
+				data = []byte(v)
+			case []byte:
+				data = v
+			}
+		}
+
+		rmsg := redisMessage{
+			message: pubsub.NewMessage{
+				Topic: stream,
+				Data:  data,
+			},
+			messageID: msg.ID,
+			handler:   handler,
+		}
 
-	return res, nil
+		select {
+		// Might block if the queue is full so we need the r.ctx.Done below.
+		case r.queue <- rmsg:
+
+		// Handle cancelation
+		case <-r.ctx.Done():
+			return
+		}
+	}
 }
 
-func (r *redisStreams) processStreams(consumerID string, streams []redis.XStream, handler func(msg *pubsub.NewMessage) error) {
-	for _, s := range streams {
-		for _, m := range s.Messages {
-			go func(stream string, message redis.XMessage) {
-				msg := pubsub.NewMessage{
-					Topic: stream,
-				}
-				data, exists := message.Values["data"]
-				if exists && data != nil {
-					msg.Data = []byte(data.(string))
-				}
+func (r *redisStreams) worker() {
+	for {
+		select {
+		// Handle cancelation
+		case <-r.ctx.Done():
+			return
 
-				err := handler(&msg)
-				if err == nil {
-					r.client.XAck(stream, consumerID, message.ID).Result()
+		case msg := <-r.queue:
+			r.logger.Debugf("Processing Redis message %s", msg.messageID)
+			if err := msg.handler(&msg.message); err == nil {
+				if err = r.client.XAck(msg.message.Topic, r.metadata.consumerID, msg.messageID).Err(); err != nil {
+					r.logger.Errorf("Error acknowledging Redis message %s: %v", msg.messageID, err)
 				}
-			}(s.Stream, m)
+			} else {
+				r.logger.Errorf("Error processing Redis message %s: %v", msg.messageID, err)
+			}
 		}
 	}
 }
 
-func (r *redisStreams) beginReadingFromStream(stream, consumerID string, handler func(msg *pubsub.NewMessage) error) {
+func (r *redisStreams) pollMessagesLoop(stream string, handler func(msg *pubsub.NewMessage) error) {
 	// first read pending items in case of recovering from crash
 	start := "0"
 
 	for {
-		streams, err := r.readFromStream(stream, consumerID, start)
+		// Read messages
+		streams, err := r.client.XReadGroup(&redis.XReadGroupArgs{
+			Group:    r.metadata.consumerID,
+			Consumer: r.metadata.consumerID,
+			Streams:  []string{stream, start},
+			Block:    0,
+		}).Result()
 		if err != nil {
 			r.logger.Errorf("redis streams: error reading from stream %s: %s", stream, err)
 
-			return
+			continue
+		}
+
+		// Enqueue messages for the returned streams
+		for _, s := range streams {
+			r.enqueue(s.Stream, consumerID, handler, s.Messages)
 		}
-		r.processStreams(consumerID, streams, handler)
 
 		// continue with new non received items
 		start = ">"
+
+		// Return on cancelation
+		if r.ctx.Err() != nil {
+			return
+		}
+	}
+}
+
+func (r *redisStreams) reclaimMessagesLoop(stream string, handler func(msg *pubsub.NewMessage) error) {
```

Comment:
I need some explanation to understand this.

Reply:
The comments I added should help